### ISSUE TITLE
RBCM

### ISSUE THREAD

**Issue by mkrompiec** [mir-group/flare#264]
Can you comment on the RobustBayesianCommitteeMachine? For what problem sizes is it faster than a "plain" GaussianProcess? What are the pitfalls? Which optimization algorithm is your 1st choice?

**Comment by nw13slx** [comment]
Hi, we do have the RBCM implemented in the development branch. 

My experience is that it is much faster for hyper-parameter optimization but not for prediction. For a training data set size of 1000, it can easily optimize the hyper-parameter within hours using 32 CPUs. Regarding the additional optimization algorithms added to the RBCM module, I haven't seen significant difference so far. In principle the basin climbing one should help with global optimization, but at least in my tests, the initial guess of hyper-parameters are still very important.

Be careful in the way of distributing training environments to the "experts". If you are handling multi-element systems, try to make sure that each expert see all elements. And environments with similar chemistry should be assigned to the same expert. I usually assign 100-200 environments to each expert and run the optimization for several time and compare the full-GP performance.

**Comment by mkrompiec** [comment]
Thank you very much for the explanation! Could you share a code example of distributing training environments to "experts"? Ideally, how to change TrajectoryTrainer to work with RBCM. The only RBCM example I've seen is the unit test which is very rudimentary...

**Comment by nw13slx** [comment]
Hi, I think @stevetorr has already implemented RBCM with TrajectoryTrainer. All you need to do is first initialize an RBCM instance
`
rbcm = RBCM(n_experts=1,  ndata_per_expert= 200, prior_variance=0.5,
                        per_expert_parallel=True, .... )
`
and then plug it as an argument to TrajectoryTrainer as if it is a normal GP.
`
tjt = TrajectoryTraine(frames, rbcm, ...)
`

Again, I have only used it for hyper-parameter training with run_passive_learning(). I'm not sure how well it performs with active learning. But please feel free to let us know if you have some functions or features in mind.

**Comment by mkrompiec** [comment]
Thanks! But I'm still struggling to get this to work. As you can see, RBCM does something strange during hyperparameter optimization (passive learning). Hyperparameters don't seem to change (or the printing function doesn't pick them up from the right place) and the likelihood goes up. 


> cat gp_from_aimd-hyps.dat
Hyperparameters: [0.443229519260524, 0.14274703191516547, 0.9689608127285437, 0.4393985495098187, 0.05]
Likelihood: -2303.290469569449
Likelihood Gradient: [-412.4097217934693, 2580.27378598003, -278.85146055128257, 968.9744931417245, -0.14632554475751736]
Hyperparameters: [0.443229519260524, 0.14274703191516547, 0.9689608127285437, 0.4393985495098187, 0.05]
Likelihood: -2324.2033267828197
Likelihood Gradient: [-435.3403135516557, 2723.2466787103326, -268.35181234355184, 920.0815753239362, -0.07079203414734933]
Hyperparameters: [0.443229519260524, 0.14274703191516547, 0.9689608127285437, 0.4393985495098187, 0.05]
Likelihood: -2376.659753131061
Likelihood Gradient: [-434.6591320408446, 2611.74136241824, -268.671457291789, 921.9756571326664, -0.05064845471665992]

**Comment by mkrompiec** [comment]
Oh, right - these are optimizations of each expert, that's why hyperparameters are the same.
